File: OsmAnd-java/src/main/java/net/osmand/ColorPalette.java
package net.osmand;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.Reader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.apache.commons.logging.Log;

public class ColorPalette {

	private static final Log LOG = PlatformUtil.getLog(ColorPalette.class);

	public static final int DARK_GREY = rgbaToDecimal(92, 92, 92, 255);
	public static final int LIGHT_GREY = rgbaToDecimal(200, 200, 200, 255);
	public static final int GREEN = rgbaToDecimal(90, 220, 95, 255);
	public static final int YELLOW = rgbaToDecimal(212, 239, 50, 255);
	public static final int RED = rgbaToDecimal(243, 55, 77, 255);
	public static final int BLUE_SLOPE = rgbaToDecimal(0, 0, 255, 255);
	public static final int CYAN_SLOPE = rgbaToDecimal(0, 255, 255, 255);
	public static final int GREEN_SLOPE = rgbaToDecimal(46, 185, 0, 255);
	public static final int WHITE = rgbaToDecimal(255, 255, 255, 255);
	public static final int YELLOW_SLOPE = rgbaToDecimal(255, 222, 2, 255);
	public static final int RED_SLOPE = rgbaToDecimal(255, 1, 1, 255);
	public static final int PURPLE_SLOPE = rgbaToDecimal(130, 1, 255, 255);

	public static final int[] COLORS = new int[] { GREEN, YELLOW, RED };
	public static final int[] SLOPE_COLORS = new int[] { CYAN_SLOPE, GREEN_SLOPE, LIGHT_GREY, YELLOW_SLOPE, RED_SLOPE };

	public static final double SLOPE_MIN_VALUE = -1.00;// -100%
	public static final double SLOPE_MAX_VALUE = 1.0;// 100%

	public static final ColorPalette SLOPE_PALETTE = parsePalette(new double[][] { { SLOPE_MIN_VALUE, BLUE_SLOPE },
			{ -0.15, CYAN_SLOPE }, { -0.05, GREEN_SLOPE }, { 0.0, LIGHT_GREY }, { 0.05, YELLOW_SLOPE },
			{ 0.15, RED_SLOPE }, { SLOPE_MAX_VALUE, PURPLE_SLOPE } });
	public static final ColorPalette MIN_MAX_PALETTE = parsePalette(
			new double[][] { { 0, GREEN }, { 0.5, YELLOW }, { 1, RED } });

	List<ColorValue> colors = new ArrayList<>();

	public ColorPalette() {
	}

	// Scale palette to min / max value from (0, 1)
	public ColorPalette(ColorPalette c, double minVal, double maxVal) {
		for (ColorValue cv : c.colors) {
			double val = cv.val * (maxVal - minVal) + minVal;
			colors.add(new ColorValue(val, cv.clr));
		}
	}

	public List<ColorValue> getColors() {
		return colors;
	}

	public static int rgbaToDecimal(int r, int g, int b, int a) {
		int value = ((a & 0xFF) << 24) | ((r & 0xFF) << 16) | ((g & 0xFF) << 8) | ((b & 0xFF) << 0);
		return value;
	}

	public int getColorByValue(double value) {
		if (Double.isNaN(value)) {
			return ColorPalette.LIGHT_GREY;
		}
		for (int i = 0; i < colors.size() - 1; i++) {
			ColorValue min = colors.get(i);
			ColorValue max = colors.get(i + 1);
			if (value == min.val)
				return min.clr;
			if (value >= min.val && value <= max.val) {
				double percent = (value - min.val) / (max.val - min.val);
				return getIntermediateColor(min, max, percent);
			}
		}
		if (value <= colors.get(0).val) {
			return colors.get(0).clr;
		} else if (value >= colors.get(colors.size() - 1).val) {
			return colors.get(colors.size() - 1).clr;
		}
		return ColorPalette.getTransparentColor();
	}

	private int getIntermediateColor(ColorValue min, ColorValue max, double percent) {
		double r = min.r + percent * (max.r - min.r);
		double g = min.g + percent * (max.g - min.g);
		double b = min.b + percent * (max.b - min.b);
		double a = min.a + percent * (max.a - min.a);
		return rgbaToDecimal((int) r, (int) g, (int) b, (int) a);
	}
	
	public static int getIntermediateColor(int min, int max, double percent) {
		double r = red(min) + percent * (red(max) - red(min));
		double g = green(min) + percent * (green(max) - green(min));
		double b = blue(min) + percent * (blue(max) - blue(min));
		double a = alpha(min) + percent * (alpha(max) - alpha(min));
		return rgbaToDecimal((int) r, (int) g, (int) b, (int) a);
	}
	
	@Override
	public String toString() {
		return writeColorPalette();
	}

	public String writeColorPalette() {
		return writeColorPalette(colors);
	}

	public static String writeColorPalette(List<ColorValue> colors) {
		StringBuilder bld = new StringBuilder();
		for (ColorValue v : colors) {
			bld.append(v.val).append(",");
			bld.append(v.r).append(",").append(v.g).append(",").append(v.b).append(",").append(v.a).append("\n");
		}
		return bld.toString().trim();
	}


	private void sortPalette() {
		Collections.sort(colors, new java.util.Comparator<ColorValue>() {
			public int compare(ColorValue a, ColorValue b) {
				return Double.compare(a.val, b.val);
			}
		});
	}

	public static class ColorValue {
		public final int r;
		public final int g;
		public final int b;
		public final int a;
		public final int clr;
		public double val;

		public ColorValue(double val, int r, int g, int b, int a) {
			this.r = r;
			this.g = g;
			this.b = b;
			this.a = a;
			this.clr = rgbaToDecimal(r, g, b, a);
			this.val = val;
		}

		public ColorValue(int clr) {
			this(0, clr);
		}

		public ColorValue(double val, int clr) {
			this.r = red(clr);
			this.g = green(clr);
			this.b = blue(clr);
			this.a = alpha(clr);
			this.clr = clr;
			this.val = val;
		}

		public void setValue(double val) {
			this.val = val;
		}

		public static ColorValue rgba(double val, int r, int g, int b, int a) {
			ColorValue clr = new ColorValue(val, r, g, b, a);
			return clr;
		}

		@Override
		public String toString() {
			return "ColorValue [r=" + r + ", g=" + g + ", b=" + b + ", a=" + a + ", val=" + val + "]";
		}

	}

	public static ColorPalette parsePalette(double[]... vl) {
		ColorPalette palette = new ColorPalette();
		for (double[] v : vl) {
			ColorValue c = null;
			if (v.length == 2) {
				c = new ColorValue(v[0], (int) v[1]);
			} else if (v.length == 4) {
				c = new ColorValue(v[0], (int) v[1], (int) v[2], (int) v[3], 255);
			} else if (v.length >= 5) {
				c = new ColorValue(v[0], (int) v[1], (int) v[2], (int) v[3], (int) v[4]);
			}
			if (c != null) {
				palette.colors.add(c);
			}
		}
		palette.sortPalette();
		return palette;
	}

	public static ColorPalette parseColorPalette(Reader reader) throws IOException {
		return parseColorPalette(reader, true);
	}

	public static ColorPalette parseColorPalette(Reader reader, boolean shouldSort) throws IOException {
		ColorPalette palette = new ColorPalette();
		BufferedReader r = new BufferedReader(reader);
		String line;
		while ((line = r.readLine()) != null) {
			String t = line.trim();
			if (t.startsWith("#")) {
				continue;
			}
			String[] values = t.split(",");
			if (values.length >= 4) {
				try {
					ColorValue rgba = ColorValue.rgba(Double.parseDouble(values[0]), Integer.parseInt(values[1]),
							Integer.parseInt(values[2]), Integer.parseInt(values[3]),
							values.length >= 5 ? Integer.parseInt(values[4]) : 255);
					palette.colors.add(rgba);
				} catch (NumberFormatException e) {
					LOG.error(e.getMessage(), e);
				}
			}
		}
		if (shouldSort) {
			palette.sortPalette();
		}
		return palette;
	}

	private static int red(int value) {
		return (value >> 16) & 0xFF;
	}

	private static int green(int value) {
		return (value >> 8) & 0xFF;
	}

	private static int blue(int value) {
		return (value >> 0) & 0xFF;
	}

	private static int alpha(int value) {
		return (value >> 24) & 0xff;
	}

	public static int getTransparentColor() {
		return rgbaToDecimal(0, 0, 0, 0);
	}

	public static void main(String[] args) throws IOException {
		ColorPalette pls = ColorPalette
				.parseColorPalette(new FileReader("../../resources/color-palette/route_speed_default.txt"));
		System.out.println(pls.colors);
		System.out.println(pls.writeColorPalette());
	}

}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/ColorPalette.kt
package net.osmand.shared

import net.osmand.shared.io.KFile
import net.osmand.shared.util.LoggerFactory
import okio.IOException
import okio.buffer

class ColorPalette {

	companion object {
		private val LOG = LoggerFactory.getLogger("ColorPalette")

		val DARK_GREY = rgbaToDecimal(92, 92, 92, 255)
		val LIGHT_GREY = rgbaToDecimal(200, 200, 200, 255)
		val GREEN = rgbaToDecimal(90, 220, 95, 255)
		val YELLOW = rgbaToDecimal(212, 239, 50, 255)
		val RED = rgbaToDecimal(243, 55, 77, 255)
		val BLUE_SLOPE = rgbaToDecimal(0, 0, 255, 255)
		val CYAN_SLOPE = rgbaToDecimal(0, 255, 255, 255)
		val GREEN_SLOPE = rgbaToDecimal(46, 185, 0, 255)
		val WHITE = rgbaToDecimal(255, 255, 255, 255)
		val YELLOW_SLOPE = rgbaToDecimal(255, 222, 2, 255)
		val RED_SLOPE = rgbaToDecimal(255, 1, 1, 255)
		val PURPLE_SLOPE = rgbaToDecimal(130, 1, 255, 255)

		val COLORS = intArrayOf(GREEN, YELLOW, RED)
		val SLOPE_COLORS = intArrayOf(CYAN_SLOPE, GREEN_SLOPE, LIGHT_GREY, YELLOW_SLOPE, RED_SLOPE)

		const val SLOPE_MIN_VALUE = -1.0
		const val SLOPE_MAX_VALUE = 1.0

		val SLOPE_PALETTE = parsePalette(
			arrayOf(
				doubleArrayOf(SLOPE_MIN_VALUE, BLUE_SLOPE.toDouble()),
				doubleArrayOf(-0.15, CYAN_SLOPE.toDouble()),
				doubleArrayOf(-0.05, GREEN_SLOPE.toDouble()),
				doubleArrayOf(0.0, LIGHT_GREY.toDouble()),
				doubleArrayOf(0.05, YELLOW_SLOPE.toDouble()),
				doubleArrayOf(0.15, RED_SLOPE.toDouble()),
				doubleArrayOf(SLOPE_MAX_VALUE, PURPLE_SLOPE.toDouble())
			)
		)

		val MIN_MAX_PALETTE = parsePalette(
			arrayOf(
				doubleArrayOf(0.0, GREEN.toDouble()),
				doubleArrayOf(0.5, YELLOW.toDouble()),
				doubleArrayOf(1.0, RED.toDouble())
			)
		)

		fun rgbaToDecimal(r: Int, g: Int, b: Int, a: Int): Int {
			return (a and 0xFF shl 24) or (r and 0xFF shl 16) or (g and 0xFF shl 8) or (b and 0xFF shl 0)
		}

		private fun parsePalette(vl: Array<DoubleArray>): ColorPalette {
			val palette = ColorPalette()
			for (v in vl) {
				val c = when {
					v.size == 2 -> ColorValue(v[0], v[1].toInt())
					v.size == 4 -> ColorValue(v[0], v[1].toInt(), v[2].toInt(), v[3].toInt(), 255)
					v.size >= 5 -> ColorValue(
						v[0],
						v[1].toInt(),
						v[2].toInt(),
						v[3].toInt(),
						v[4].toInt()
					)

					else -> null
				}
				c?.let { palette.colors.add(it) }
			}
			palette.sortPalette()
			return palette
		}

		fun getTransparentColor(): Int {
			return rgbaToDecimal(0, 0, 0, 0)
		}

		private fun red(value: Int): Int {
			return (value shr 16) and 0xFF
		}

		private fun green(value: Int): Int {
			return (value shr 8) and 0xFF
		}

		private fun blue(value: Int): Int {
			return value and 0xFF
		}

		private fun alpha(value: Int): Int {
			return value shr 24 and 0xFF
		}

		fun getIntermediateColor(min: Int, max: Int, percent: Double): Int {
			val r: Double = red(min) + percent * (red(max) - red(min))
			val g: Double = green(min) + percent * (green(max) - green(min))
			val b: Double = blue(min) + percent * (blue(max) - blue(min))
			val a: Double = alpha(min) + percent * (alpha(max) - alpha(min))
			return rgbaToDecimal(r.toInt(), g.toInt(), b.toInt(), a.toInt())
		}
		
		@Throws(IOException::class)
		fun parseColorPalette(file: KFile): ColorPalette {
			val palette = ColorPalette()
			val buf = file.source().buffer()
			var line: String?
			while (true) {
				line = buf.readUtf8Line()
				if (line == null) break
				val t = line.trim { it <= ' ' }
				if (t.startsWith("#")) {
					continue
				}
				val values = t.split(",".toRegex()).dropLastWhile { it.isEmpty() }
					.toTypedArray()
				if (values.size >= 4) {
					try {
						val rgba: ColorValue =
							ColorValue.rgba(
								values[0].toDouble(),
								values[1].toInt(),
								values[2].toInt(),
								values[3].toInt(),
								if (values.size >= 4) values[4].toInt() else 255
							)
						palette.colors.add(rgba)
					} catch (e: NumberFormatException) {
						LOG.error(e.message, e)
					}
				}
			}
			palette.sortPalette()
			return palette
		}
	}

	val colors = mutableListOf<ColorValue>()

	constructor()

	constructor(c: ColorPalette, minVal: Double, maxVal: Double) {
		for (cv in c.colors) {
			val value = cv.value * (maxVal - minVal) + minVal
			colors.add(ColorValue(value, cv.clr))
		}
	}

	fun getColorByValue(value: Double): Int {
		if (value.isNaN()) {
			return LIGHT_GREY
		}
		for (i in 0 until colors.size - 1) {
			val min = colors[i]
			val max = colors[i + 1]
			if (value == min.value) return min.clr
			if (value >= min.value && value <= max.value) {
				val percent = (value - min.value) / (max.value - min.value)
				return getIntermediateColor(min, max, percent)
			}
		}
		return when {
			value <= colors[0].value -> colors[0].clr
			value >= colors[colors.size - 1].value -> colors[colors.size - 1].clr
			else -> getTransparentColor()
		}
	}

	private fun getIntermediateColor(min: ColorValue, max: ColorValue, percent: Double): Int {
		val r = min.r + percent * (max.r - min.r)
		val g = min.g + percent * (max.g - min.g)
		val b = min.b + percent * (max.b - min.b)
		val a = min.a + percent * (max.a - min.a)
		return rgbaToDecimal(r.toInt(), g.toInt(), b.toInt(), a.toInt())
	}

	override fun toString(): String {
		return writeColorPalette()
	}

	fun writeColorPalette(): String {
		val bld = StringBuilder()
		for (v in colors) {
			bld.append(v.value).append(",")
			bld.append(v.r).append(",").append(v.g).append(",").append(v.b).append(",").append(v.a)
				.append("\n")
		}
		return bld.toString().trim()
	}

	private fun sortPalette() {
		colors.sortWith(compareBy { it.value })
	}

	class ColorValue(
		val value: Double,
		val r: Int,
		val g: Int,
		val b: Int,
		val a: Int,
		val clr: Int = rgbaToDecimal(r, g, b, a)
	) {
		constructor(value: Double, clr: Int) : this(
			value,
			red(clr),
			green(clr),
			blue(clr),
			alpha(clr),
			clr
		)

		override fun toString(): String {
			return "ColorValue [r=$r, g=$g, b=$b, a=$a, val=$value]"
		}

		companion object {
			fun rgba(value: Double, r: Int, g: Int, b: Int, a: Int): ColorValue {
				return ColorValue(value, r, g, b, a)
			}
		}
	}
}
